pkg/chunk/cached_store.go

```diff
@@ -965,10 +966,19 @@ func (store *cachedStore) uploadStagingFile(key string, stagingPath string) {
 		logger.Errorf("Read staging file %s: %s", stagingPath, err)
 		return
 	}
+	if !store.isPendingValid(key) {
+		block.Release()
+		logger.Debugf("Key %s is not needed, drop it", key)
+		return
+	}
 
 	store.stageBlockDelay.Add(time.Since(item.ts).Seconds())
 	if err = store.upload(key, block, nil); err == nil {
 		store.bcache.uploaded(key, blen)
+		if !store.isPendingValid(key) { // Delete leaked objects if it's already deleted by other goroutines
```

Comment:
Skip other functions (`uploaded`, `removePending`, `removeStage`) when isPendingValid = false.

Reply:
Is it safe? I suppose at least we still need `removePending`?

Reply:
pending not valid means it's already removed from the pending dict

Reply:
Right, updated